inst/include/mode/r/mode.hpp

```diff
@@ -85,29 +85,41 @@ cpp11::sexp mode_stats(SEXP ptr) {
 }
 
 template <typename T>
-void mode_update_state(SEXP ptr, SEXP r_pars, SEXP r_state, SEXP r_time,
+void mode_update_state(SEXP ptr, SEXP r_pars, SEXP r_time, SEXP r_state,
+                       SEXP r_index,
                        SEXP r_set_initial_state,
                        SEXP r_reset_step_size) {
   mode::container<T> *obj =
       cpp11::as_cpp < cpp11::external_pointer<mode::container<T>>>(ptr).get();
 
+  std::vector<size_t> index;
+  const size_t index_max = obj->n_state_full();
+  if (r_index != R_NilValue) {
+    index = mode::r::r_index_to_index(r_index, index_max);
+  } else {
+    index.clear();
+    index.reserve(index_max);
+    for (size_t i = 0; i < index_max; ++i) {
+      index.push_back(i);
```

Comment:
seemed easiest to just build a full index if the user hasn't provided one, to avoid further complicating the logic in the internal function